packages/x-tree-view/src/internals/TreeViewProvider/TreeViewChildrenItemProvider.tsx

```diff
@@ -0,0 +1,125 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import { useTreeViewContext } from './useTreeViewContext';
+import type { UseTreeViewJSXNodesSignature } from '../plugins/useTreeViewJSXNodes';
+import { TreeViewItemChildrenIndexes } from '../plugins/useTreeViewNodes/useTreeViewNodes.types';
+
+function binaryFindPosition(
+  otherDescendants: { id: string; element: HTMLLIElement }[],
+  element: HTMLLIElement,
+) {
+  let start = 0;
+  let end = otherDescendants.length - 1;
+
+  while (start <= end) {
+    const middle = Math.floor((start + end) / 2);
+
+    if (
+      // eslint-disable-next-line no-bitwise
+      otherDescendants[middle].element.compareDocumentPosition(element) &
+      Node.DOCUMENT_POSITION_PRECEDING
+    ) {
+      end = middle - 1;
+    } else {
+      start = middle + 1;
+    }
+  }
+
+  return start;
+}
+
+export const TreeViewChildrenItemContext =
+  React.createContext<TreeViewChildrenItemContextValue | null>(null);
+
+if (process.env.NODE_ENV !== 'production') {
+  TreeViewChildrenItemContext.displayName = 'TreeViewChildrenItemContext';
+}
+
+interface TreeViewChildrenItemProviderProps {
+  id?: string;
+  rootRef?: React.RefObject<HTMLElement>;
+  children: React.ReactNode;
+}
+
+export function TreeViewChildrenItemProvider(props: TreeViewChildrenItemProviderProps) {
+  const { children, rootRef, id = null } = props;
+
+  const { instance } = useTreeViewContext<[UseTreeViewJSXNodesSignature]>();
+  const parentContext = React.useContext(TreeViewChildrenItemContext);
+  const childrenElementRef = React.useRef<Map<string, HTMLLIElement>>(new Map());
+  const cachedDescendantIds = React.useRef('');
+
+  React.useEffect(() => {
+    let element: HTMLElement | null | undefined;
+    if (rootRef) {
+      element = rootRef.current;
+    } else if (id && parentContext) {
+      // The DescendantProvider is not aware of the ref of its own item,
+      // but its parent can give him this information.
+      element = parentContext.getChild(id);
+    }
+
+    // Weak but quick way to check if the descendants have changed and avoid the expensive DOM operations.
+    // This is not sufficient to get the new order because this includes grand-children, not only direct children.
+    if (element) {
+      const newDescendantIds = Array.from(element.querySelectorAll('[role="treeitem"]'))
```

Comment:
A couple of possibly nitpick topics:
1. Have you considered treating the DOM as the source of truth and adjusting event handling to interact with DOM rather than a tree item indexed representation that we try to keep in sync?
2. I noticed that in our implementation only `treeitem`s of an `expanded` `group` are rendered. Was this intentional for some reason? 🤔 
I think it is somewhat of an a11y problem because screen readers can not give context about what is inside of a collapsed group. 🙈

Reply:
> Have you considered treating the DOM as the source of truth and adjusting event handling to interact with DOM rather than a tree item indexed representation that we try to keep in sync?

Yes
And the challenge is to have `RichTreeView` working with it since it has the index without looking on the DOM.
Should we have 2 ways of getting the index (`RichTreeView` uses state and `SimpleTreeView` looks on the DOM), or should we always look on the DOM even though `RichTreeView` has the information in state?
Always looking on the DOM will probably not work well with virtualization (which is only planned for `RichTreeView`).

>  noticed that in our implementation only treeitems of an expanded group are rendered. Was this intentional for some reason?

Do we agree that it was already the case before this PR?
I think it's a performance topic mostly, it was like this when I migrated but your point about a11y is interesting...

Reply:
> And the challenge is to have RichTreeView working with it since it has the index without looking on the DOM.
Should we have 2 ways of getting the index (RichTreeView uses state and SimpleTreeView looks on the DOM), or should we always look on the DOM even though RichTreeView has the information in state?
Always looking on the DOM will probably not work well with virtualization (which is only planned for RichTreeView).

Yes, I do understand that `RichTreeView` couldn't work with DOM event management, specifically because of virtualization and it does not need, because we already have one source of truth in that case.
I was indeed referring to the consideration of going with a simpler approach for the `SimpleTreeView`, albeit, it would be different between the components. 🤔 

In any case, it can probably be tackled without a BC if we see that as the better option, because it's mostly internal logic. 🤷 

> Do we agree that it was already the case before this PR?
I think it's a performance topic mostly, it was like this when I migrated but your point about a11y is interesting...

Oh yes, I just mentioned it here, that's not related to this PR.
I just double-checked the a11y behavior and it seems that I might have "dreamt" a behavioral difference initially.
Even the [tree view with all items rendered](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/examples/treeview-1a/) does not announce anything about children until the group/parent is expanded. 👌 
So, there does not seem to be any direct a11y consequence because of this. 👌

Reply:
> I just double-checked the a11y behavior and it seems that I might have "dreamt" a behavioral difference initially.
Even the [tree view with all items rendered](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/examples/treeview-1a/) does not announce anything about children until the group/parent is expanded. 👌

:ok_hand: That's reassuring